pail/s3_bucket.go

```diff
@@ -0,0 +1,551 @@
+package pail
+
+import (
+	"context"
+	"io"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/s3"
+	"github.com/pkg/errors"
+)
+
+type s3BucketSmall struct {
+	s3Bucket
+}
+
+type s3BucketLarge struct {
+	s3Bucket
+}
+
+type s3Bucket struct {
+	name   string
+	prefix string
+	sess   *session.Session
+	svc    *s3.S3
+}
+
+type S3Options struct {
+	// If do not provide credentials, fall back to AWS default provider chain.
+	Credentials *credentials.Credentials
+	Region      string
+	Name        string
+	Prefix      string
+	Access      string // for puts only, presumably.
+}
+
+func (s *s3Bucket) normalizeKey(key string) string {
+	if key == "" {
+		return s.prefix
+	}
+	if s.prefix != "" {
+		return filepath.Join(s.prefix, key)
+	}
+	return key
+}
+
+func (s *s3Bucket) denormalizeKey(key string) string {
+	if s.prefix != "" {
+		denormalizedKey, err := filepath.Rel(s.prefix, key)
+		if err != nil {
+			return key
+		}
+		return denormalizedKey
+	}
+	return key
+}
+
+func newS3Bucket(s3Options S3Options) (*s3Bucket, error) {
+	config := &aws.Config{Region: aws.String(s3Options.Region)}
+	if s3Options.Credentials != nil {
+		_, err := s3Options.Credentials.Get()
+		if err != nil {
+			return &s3Bucket{}, errors.Wrap(err, "invalid credentials!")
+		}
+		config.Credentials = s3Options.Credentials
+	}
+	sess, err := session.NewSession(config)
+	if err != nil {
+		return &s3Bucket{}, errors.Wrap(err, "problem connecting to AWS")
+	}
+	svc := s3.New(sess)
+	return &s3Bucket{name: s3Options.Name, prefix: s3Options.Prefix, sess: sess, svc: svc}, nil
+}
+
+func NewS3BucketSmall(s3Options S3Options) (Bucket, error) {
```

Comment:
can these functions have docstrings, and perhaps name them "S3Bucket" and "S3MultiPartBucket"